app/services/foreman/deprecation.rb

```diff
@@ -0,0 +1,19 @@
+module Foreman
+  class Deprecation
+    #aviliable options: controller, method, info, callstack
+    #deadline_version is the last version the method could be used
+    def self.deprecation_warning(foreman_version_deadline, options = {})
+      raise Foreman::Exception, N_("Invalid version format, please enter in x.y.z") unless foreman_version_deadline.to_s.match(/\A\d[.]\d+[.]\d\z/)
+      callstack = options[:callstack] || caller
+      log_msg = "You are using a deprecated behavior, #{foreman_version_deadline} is the last version it will be available for use."
+      log_msg += " Use #{options[:controller]}##{options[:method]} instead." if options[:controller] && options[:method]
+      log_msg += " #{options[:info]}." if options[:info]
+      ActiveSupport::Deprecation.warn(log_msg, callstack)
+      raise Foreman::Exception, N_("You are using a deprecated behavior, for more information check log file. %{callstack}") % {:callstack => ActiveSupport::Deprecation.send(:deprecation_caller_message, callstack)} if version_is_outdated?(foreman_version_deadline)
```

Comment:
1. I wanted to give more info about the exception source, it was already implemented on ActiveSupport::Deprecation.deprecation_caller_message.(without sending the callstack the source will always be the deprecation class)
I can  take it off and just direct to the log for more info, what do you think?